### PR TITLE
Refactor term registration visitors

### PR DESCRIPTION
**Pull request by ajreynol**
This refactors the term registration visitors (PreRegisterVisitor and SharedTermsVisitor), which are responsible for calling Theory::preRegisterTerm and Theory::addSharedTerm.  

The purpose of this PR is to resolve a subtle bug in the implementation of PreRegisterVisitor (see explanation below), and to improve performance in particular on the Facebook benchmarks (where preregistering terms is currently 25-35% of runtime on certain challenge benchmarks).

This PR makes it so that that SharedTermsVisitor now subsumes PreRegisterVisitor and should be run when sharing is enabled *only*.  Previously, we ran *both* PreRegisterVisitor and SharedTermsVisitor, and moreover the former misclassified when a literal had multiple theories since the `d_theories` field of PreRegisterVisitor is never reset.  This meant we would consider a literal to have multiple theories as soon as we saw *any* literal that had multipleTheories.  After this PR, we run SharedTermsVisitor only which also handles calling `preRegisterTerm`, since it has the ability to anyways (it computes the same information as PreRegisterVisitor regarding which theories care about a term).  

The refactoring merges the blocks of code that were copy and pasted in term_registration_visitor.cpp and makes them more optimal, by reducing our calls to Theory::theoryOf.

FYI @barrettcw

### CODE REVIEW

File: src/theory/term_registration_visitor.cpp [hunk 1]
```diff
@@ -32,6 +32,46 @@ std::string PreRegisterVisitor::toString() const {
   return ss.str();
 }
 
+/**
+ * Return true if we should visit current from parent, given that the
+ * set of theories in visitedTheories has already visited current. This method
+ * is used by both visitors below.
```
Comment (yoni206) [inline src/theory/term_registration_visitor.cpp:38]: I suggest explicitly writing the names of both visitors.

File: src/theory/term_registration_visitor.cpp [hunk 2]
```diff
@@ -47,62 +87,16 @@ bool PreRegisterVisitor::alreadyVisited(TNode current, TNode parent) {
...
-      } else {
-        return true;
-      }
-    } else {
-      return false;
-    }
-  } else {
-    return false;
-  }
+  return !shouldVisit(visitedTheories, current, parent);
```
Comment (yoni206) [inline src/theory/term_registration_visitor.cpp:99]: It is a bit confusing to me that "alreadyVisited" depends on "shouldVisit". I would expect it to be the other way around. Can you clarify this in a commend, namely -- explain the call the `shouldVisit`?

File: src/theory/term_registration_visitor.cpp [hunk 3]
```diff
@@ -112,74 +106,96 @@ void PreRegisterVisitor::visit(TNode current, TNode parent) {
...
-
-  Assert(d_visited.find(current) != d_visited.end());
-  Assert(alreadyVisited(current, parent));
+  // call the theory's preRegisterTerm method
+  visitedTheories = TheoryIdSetUtil::setInsert(id, visitedTheories);
+  Theory* th = te->theoryOf(id);
+  th->preRegisterTerm(current);
 }
 
+void PreRegisterVisitor::start(TNode node) {}
```
Comment (yoni206) [inline src/theory/term_registration_visitor.cpp:197]: what is this for?
If this is supposed to be overriden always, maybe assert false?

File: src/theory/term_registration_visitor.h [hunk 4]
```diff
@@ -50,22 +50,17 @@ class PreRegisterVisitor {
...
-
   /**
    * String representation of the visited map, for debugging purposes.
    */
   std::string toString() const;
 
  public:
   /** Returned set tells us which theories there are */
-  typedef theory::TheoryIdSet return_type;
+  typedef void return_type;
```
Comment (yoni206) [inline src/theory/term_registration_visitor.h:60]: Can you clarify this?

**Review by yoni206 — COMMENTED** [review]
LGTM except for a few questions/comments below.
If @barrettcw can also review, that would be great, as this part of the code might be risky for combination with finite types.

File: src/theory/term_registration_visitor.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/term_registration_visitor.cpp:197]: Its a slightly different style, this visitor is intended to instantiate a template type on NodeVisitor (https://github.com/CVC4/CVC4/blob/master/src/expr/node_visitor.h), which is why these methods need to be defined.

I'm not a big fan of this style, but I don't feel strongly about refactoring this further (yet).

File: src/theory/term_registration_visitor.h (hunk 4 above)
Comment (ajreynol) [reply src/theory/term_registration_visitor.h:60]: See above.

File: src/theory/term_registration_visitor.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/term_registration_visitor.cpp]: I renamed the method to `isAlreadyVisited`.

**Review by barrettcw — COMMENTED** [review]
The changes are pretty subtle and I won't be able to check them carefully.  I suggest you run regressions + some subset of SMT-LIB to ensure nothing is broken.

**Comment by ajreynol** [comment]
Results on QF benchmarks from SMT-LIB with combined theories (without FP), 60 sec timeout:

```
                           Configuration  #unsat      time    #sat      time #solved    #total
                       CVC4-020921-rtr_d    6795   8887.97   12641     20139     19436     20725
                           CVC4-020921_d    6805   9474.34   12641   20028.3     19446     20725



For unsatisfiable : 
                           CVC4-020921_d       X      26
                       CVC4-020921-rtr_d      16       X

For satisfiable : 
                           CVC4-020921_d       X       1
                       CVC4-020921-rtr_d       1       X

```
Benchmarks that are different status are mostly close to timeout:
|                                                                                | NEW   |         | OLD   |         |
| ------------------------------------------------------------------------------ | ----- | ------- | ----- | ------- |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000255.smt2     | t/o   | 60.0236 | unsat | 58.7769 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000206.smt2     | t/o   | 60.097  | unsat | 58.2942 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000237.smt2     | t/o   | 60.1037 | unsat | 58.3451 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000041.smt2     | unsat | 59.8598 | t/o   | 60.1083 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000233.smt2     | t/o   | 60.0147 | unsat | 58.9732 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000202.smt2     | t/o   | 60.023  | unsat | 59.6719 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000216.smt2     | t/o   | 60.0219 | unsat | 59.6534 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000049.smt2     | t/o   | 60.1094 | unsat | 59.9864 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000061.smt2     | t/o   | 60.0653 | unsat | 59.0428 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000280.smt2     | unsat | 59.3821 | t/o   | 60.0177 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000048.smt2     | t/o   | 60.0223 | unsat | 59.5981 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000167.smt2     | unsat | 59.3655 | t/o   | 60.107  |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000269.smt2     | t/o   | 60.0231 | unsat | 59.0263 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000090.smt2     | t/o   | 60.0322 | unsat | 59.915  |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000156.smt2     | t/o   | 60.0828 | unsat | 58.7199 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000017.smt2     | unsat | 59.3494 | t/o   | 60.0221 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000071.smt2     | unsat | 59.6584 | t/o   | 60.0212 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000248.smt2     | t/o   | 60.0175 | unsat | 58.1743 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000246.smt2     | unsat | 59.2506 | t/o   | 60.0142 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000179.smt2     | t/o   | 60.016  | unsat | 59.0455 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000277.smt2     | unsat | 59.8955 | t/o   | 60.0227 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000005.smt2     | t/o   | 60.0252 | unsat | 58.8394 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000030.smt2     | t/o   | 60.0948 | unsat | 59.2827 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000113.smt2     | t/o   | 60.0979 | unsat | 59.227  |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000077.smt2     | unsat | 59.7375 | t/o   | 60.02   |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000009.smt2     | unsat | 59.5373 | t/o   | 60.072  |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000126.smt2     | t/o   | 60.0166 | unsat | 59.2582 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000038.smt2     | t/o   | 60.0142 | unsat | 59.639  |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000219.smt2     | unsat | 59.8379 | t/o   | 60.0313 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000230.smt2     | t/o   | 60.0917 | unsat | 58.4096 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000082.smt2     | t/o   | 60.0445 | unsat | 58.7974 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000170.smt2     | unsat | 59.6264 | t/o   | 60.0609 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000235.smt2     | t/o   | 60.0962 | unsat | 59.9327 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000102.smt2     | t/o   | 60.0225 | unsat | 59.5773 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000191.smt2     | unsat | 59.418  | t/o   | 60.0326 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000276.smt2     | t/o   | 60.015  | unsat | 59.0889 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000214.smt2     | unsat | 59.2808 | t/o   | 60.0964 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000004.smt2     | t/o   | 60.056  | unsat | 58.9771 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000272.smt2     | t/o   | 60.0447 | unsat | 59.4971 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000195.smt2     | unsat | 59.2721 | t/o   | 60.0148 |
| 2020-tc/QF\_ABV/klee-selected-smt2/cu-no-caches/\_csplit-query-000208.smt2     | unsat | 59.1838 | t/o   | 60.0149 |
| 2020-tc/QF\_ABV/platania/no\_init\_multi\_delete/no\_init\_multi\_delete8.smt2 | t/o   | 60.0254 | sat   | 54.8466 |
| 2020-tc/QF\_ABV/stp/testcase8.stp.smt2                                         | unsat | 59.8693 | t/o   | 60.0232 |
| 2020-tc/QF\_ALIA/qlock2/qlock.base.29.smt2                                     | sat   | 26.6839 | t/o   | 60.0901 |

**Review by barrettcw — APPROVED** [review]
OK LGTM

**Pull request merged**